sycl/include/sycl/nd_item.hpp

```diff
@@ -38,6 +41,492 @@ namespace ext::oneapi::experimental {
 template <int Dimensions> class root_group;
 }
 
+#if __INTEL_PREVIEW_BREAKING_CHANGES
+/// Identifies an instance of the function object executing at each point in an
+/// nd_range.
+///
+/// \ingroup sycl_api
+template <int Dimensions = 1> class nd_item {
+public:
+  static constexpr int dimensions = Dimensions;
+
+  id<Dimensions> get_global_id() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalInvocationId<Dimensions, id<Dimensions>>();
+#else
+    assert(false && "nd_item methods can't be used on the host!");
```

Comment:
Should we throw instead? Currently, a user running non-debug code would just receive the wrong result without any error/warning, and it might be confusing. 

Same for other locations were we do something similar.

Reply:
what about something like this? This compiles; whether it should/has defined behavior is unclear - should nd_item be device copyable?

```c++
#include <sycl/sycl.hpp>

int main() {
  using namespace sycl;
  static_assert(is_device_copyable_v<nd_item<1>>);
  static_assert(is_device_copyable_v<nd_item<2>>);
  static_assert(is_device_copyable_v<nd_item<3>>);
  queue q;
  {
    auto *p = malloc_shared<nd_item<1>>(1, q);
    q.parallel_for(nd_range<1>(8, 4), [=](nd_item<1> it) {
      if (it.get_global_id(0) == 5)
        *p = it;
    }).wait();
    if (p->get_global_id(0) != 5) std::cout << "fail\n";
    free(p, q);
  }
  {
    buffer<nd_item<1>, 1> buf { 1 };
    q.submit([&](handler &cgh) {
      accessor acc { buf, cgh };
      cgh.parallel_for(nd_range<1>(8,4), [=](nd_item<1> it) {
        if (it.get_global_id(0) == 5)
          acc[0] = it;
      });
    });
    host_accessor ha { buf };
    if (ha[0].get_global_id(0) != 5) std::cout << "fail\n";
  }
}
```

Reply:
This should not be valid. The spec doesn't say it explicitly, but `nd_item` should not be available on the host.

https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_class_availability section says that it must be available in SYCL kernels.
https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:byval-semantics assumes that some of these types are not available on the host.

`nd_item` must be one the classes which are not available on the host. `nd_item` methods have meaning only in SYCL kernel scope.

@intel/dpcpp-specification-reviewers, please, follow-up on spec clarifications if needed.

Reply:
> `nd_item` must be one the classes which are not available on the host. `nd_item` methods have meaning only in SYCL kernel scope.

Exactly, totally agreed. That's why I think throwing would clearly state that this is not an implementation error, but a user error, if anybody is not very familiar with SYCL and tries to do something like Justin's example.

Anyway, this is not blocking and I can accept either way.

Reply:
Done in https://github.com/intel/llvm/pull/12236/commits/051ce6c68b9d4f5441eff36f9d61fbe7b20e6ae0.